server/src/queries.rs

```diff
@@ -0,0 +1,252 @@
+            flight_request.arrival_time.as_ref().unwrap().seconds,
+            flight_request.arrival_time.as_ref().unwrap().nanos as u32,
+        ));
+        (
+            arrival_time - Duration::minutes(block_aircraft_minutes as i64),
+            arrival_time,
+        )
+    };
+    let is_departure_vertiport_available = departure_vertiport_schedule.is_available_between(
+        departure_time,
+        departure_time + Duration::minutes(block_departure_vertiport_minutes as i64),
+    );
+    let departure_vertiport_flights: Vec<FlightPlan> = vec![];
+    /*todo storage_client
+    .flight_plans(Request::new(FlightPlanFilter {})) //todo filter flight_plans(estimated_departure_between: ($from, $to), vertiport_id: $ID)
+    .await?
+    .into_inner()
+    .flight_plans;*/
+
+    if !is_departure_vertiport_available || !departure_vertiport_flights.is_empty() {
+        return Err(Status::not_found("Departure vertiport not available"));
+    }
+
+    let is_arrival_vertiport_available = arrival_vertiport_schedule.is_available_between(
+        arrival_time - Duration::minutes(block_arrival_vertiport_minutes as i64),
+        arrival_time,
+    );
+    if !is_arrival_vertiport_available {
+        return Err(Status::not_found("Arrival vertiport not available"));
+    }
+    let arrival_vertiport_flights: Vec<FlightPlan> = vec![];
+    /* todo storage_client
+    .flight_plans(Request::new(FlightPlanFilter {}))
+    //todo filter flight_plans(estimated_arrival_between: ($from, $to), vertiport_id: $ID)
+    .await?
+    .into_inner()
+    .flight_plans;*/
+    if !is_arrival_vertiport_available || !arrival_vertiport_flights.is_empty() {
+        return Err(Status::not_found("Arrival vertiport not available"));
+    }
+    //5. check schedule of aircrafts
+    let aircrafts = storage_client
+        .aircrafts(Request::new(AircraftFilter {})) //todo filter associated aircrafts to dep vertiport?
+        .await?
+        .into_inner()
+        .aircrafts;
+    for _aircraft in aircrafts {
+        let aircraft_schedule = Calendar::from_str(SAMPLE_CAL).unwrap(); //TODO get from aircraft.schedule
+        let is_aircraft_available =
+            aircraft_schedule.is_available_between(departure_time, arrival_time);
+        let aircraft_flights: Vec<FlightPlan> = vec![];
+        /*todo storage_client
+        .flight_plans(Request::new(FlightPlanFilter {}))
+        //todo filter flight_plans(estimated_departure_between: ($from, $to), estimated_arrival_between: ($from2, $to2) aircraft_id: $ID)
+        .await?
+        .into_inner()
+        .flight_plans;*/
+        if !is_aircraft_available || !aircraft_flights.is_empty() {
+            return Err(Status::not_found("Aircraft not available"));
+        }
+    }
+    //6. TODO: check other constraints (cargo weight, number of passenger seats)
+    //7. create draft flight plan (in memory)
+    let fp_id = Uuid::new_v4().to_string();
```

Comment:
Please comment a TODO to compare against other fp UUIDs to ensure no duplicates

This would be a call to svc-storage

The likelihood is extremely low but not zero

R2 requirement

Reply:
If we keep draft flight plans only in scheduler, then it is necessary to generate some id locally. My original implementation was to create drafts in svc-storage and then update status do confirmed. But we changed it.
This Id is only for the local hashmap, when we insert the record in to database, DB will create the real Id.
Obtaining and reserving Id from storage is basically the same as creating a draft.

Reply:
Ah okay. That makes sense, and we don't end up reserving/wasting unique identifiers on draft plans when they're discarded. Sounds good to me